### PR TITLE
PLAT-7018 Allow rewrite batches for replace and duplicate statements

### PR DESCRIPTION
**Pull request by oyeliseiev-ua**
- Allow rewrite batch command for 'REPLACE' and 'ON DUPLICATE KEY'
- Fix rewrite batch command for not parameterized queries or partially parameterized
- In case if rewriteBatchedStatements is not possible try to execute in multi queries command
- Add more accurate packet size calculation to split by maxAllowedPacket

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -618,7 +618,6 @@
                     <groupId>ch.qos.logback</groupId>
                     <artifactId>logback-classic</artifactId>
                     <version>${logback.version}</version>
-                    <scope>test</scope>
```
Comment (AdalbertMemSQL) [inline pom.xml:621]: Why did you change this?
Is there any functionality in the driver that requires logback?

File: src/main/java/com/singlestore/jdbc/client/impl/StandardClient.java [hunk 2]
```diff
@@ -406,7 +406,7 @@ public int sendQuery(ClientMessage message) throws SQLException {
       if (timeOut) {
         throw exceptionFactory
             .withSql(message.description())
-            .create("Socket error: query timed out", "08000", ioException);
+            .create("Query execution was interrupted", "70100", ioException);
```
Comment (AdalbertMemSQL) [inline src/main/java/com/singlestore/jdbc/client/impl/StandardClient.java:409]: Nice change!
Can we also mention the parameter which controls the timeout?

File: pom.xml (hunk 1 above)
Comment (oyeliseiev-ua) [reply pom.xml:621]: It's in **bench** profile used for performance runs

**Review by pmishchenko-ua — APPROVED** [review]

File: src/main/java/com/singlestore/jdbc/plugin/codec/ByteArrayCodec.java [hunk 3]
```diff
@@ -49,6 +50,11 @@ public boolean canEncode(Object value) {
     return value instanceof byte[];
   }
 
+  @Override
+  public int getApproximateTextProtocolLength(Object value) throws SQLException {
+    return canEncode(value) ? ((byte[]) value).length * 2 : -1;
```
Comment (AdalbertMemSQL) [inline src/main/java/com/singlestore/jdbc/plugin/codec/ByteArrayCodec.java:55]: Why do we multiply it here?

File: src/main/java/com/singlestore/jdbc/plugin/codec/StringCodec.java [hunk 4]
```diff
@@ -64,6 +65,11 @@ public boolean canEncode(Object value) {
     return value instanceof String;
   }
 
+  @Override
+  public int getApproximateTextProtocolLength(Object value) throws SQLException {
+    return canEncode(value) ? String.valueOf(value).length() * 3 : -1;
```
Comment (AdalbertMemSQL) [inline src/main/java/com/singlestore/jdbc/plugin/codec/StringCodec.java:70]: Why do we multiply by 3 here?

File: src/main/java/com/singlestore/jdbc/util/ClientParser.java [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: LGPL-2.1-or-later
 // Copyright (c) 2012-2014 Monty Program Ab
-// Copyright (c) 2015-2023 MariaDB Corporation Ab
-// Copyright (c) 2021-2023 SingleStore, Inc.
+// Copyright (c) 2015-2024 MariaDB Corporation Ab
+// Copyright (c) 2021-2024 SingleStore, Inc.
```
Comment (AdalbertMemSQL) [inline src/main/java/com/singlestore/jdbc/util/ClientParser.java:4]: I think that we can change it to 
```
// Copyright (c) 2015-2021 MariaDB Corporation Ab
// Copyright (c) 2021 SingleStore, Inc.
```
Otherwise we will need to change it every year :)

File: src/main/java/com/singlestore/jdbc/util/RewriteClientParser.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    this.queryPartsLength = queryPartsLength;
+    this.isQueryMultiValuesRewritable = isQueryMultiValuesRewritable;
+    this.paramCount = queryParts.size() - 3;
+  }
+
+  /**
+   * Separate query in a String list and set flag isQueryMultiValuesRewritable The parameters "?"
+   * (not in comments) emplacements are to be known.
+   *
+   * <p>The only rewritten queries follow these notation: INSERT [LOW_PRIORITY | DELAYED |
```
Comment (AdalbertMemSQL) [inline src/main/java/com/singlestore/jdbc/util/RewriteClientParser.java:37]: SingleStore don't have [LOW_PRIORITY | DELAYED | HIGH_PRIORITY]

**Review by AdalbertMemSQL — APPROVED** [review]
I left several minor comments but in general, looks ok.

File: src/main/java/com/singlestore/jdbc/plugin/codec/StringCodec.java (hunk 4 above)
Comment (oyeliseiev-ua) [reply src/main/java/com/singlestore/jdbc/plugin/codec/StringCodec.java:70]: is max size possible: 1 char = 2 bytes + escape character = 1 byte encoded (method encode -> writeStringEscaped)

File: src/main/java/com/singlestore/jdbc/plugin/codec/ByteArrayCodec.java (hunk 3 above)
Comment (oyeliseiev-ua) [reply src/main/java/com/singlestore/jdbc/plugin/codec/ByteArrayCodec.java:55]: looking to encoding method writeBytesEscaped - it required 2*length buffer when adding escape slash
